fcos|r4e|flatcar spec stabilization (#442)
coreos/butane: docs/upgrading-fcos.md

Reviewer: The existing draft release notes for the new specs are in `{% comment %}` blocks and will need to be uncommented.

Author: Oh, ok. I thought that would happen during the actual release?

Reviewer: The spec stabilization is happening in this PR, so this PR is the right place to document the changes.

Author: Ok I dropped the comment blocks.

1. TOC
{:toc}

## From Version 1.4.0 to Version 1.5.0

There are no breaking changes between versions 1.4.0 and 1.5.0 of the `fcos` configuration specification. Any valid 1.4.0 configuration can be updated to a 1.5.0 configuration by changing the version string in the config.

The following is a list of notable new features.

> ## GRUB password

Reviewer: Maybe `GRUB passwords`?  And this should be a level-3 heading `###` rather than level 2.

> A new `grub` section was added with an array of users can be set with `name`(s) and `password_hash`(s). Any user that does not have the given usernames and passwords will not be able to use the GRUB command line, modify kernel command-line arguments, or boot non-default OSTree deployments.

Reviewer: Maybe:

```suggestion
The config gained a new top-level `grub` section. It contains a `users` section with a list of usernames and corresponding password hashes for authenticating to the GRUB bootloader. If any users are specified, GRUB will require authentication before using the GRUB command line, modifying kernel command-line arguments, or booting non-default OSTree deployments. Password hashes can be generated with `grub2-mkpasswd-pbkdf2`.
```

We should also document the command elsewhere; PR in #445.

Author: Thats much better, thank you for the wording fixes.

<!-- butane-config -->
```yaml
variant: fcos
version: 1.5.0
grub:
  users:

>     - name: root

Reviewer: Let's use an arbitrary name like `admin` to make it clear that this isn't the same as the system's root password.

      password_hash: grub.pbkdf2.sha512.10000.874A958E5264...
```

### Offline Tang provisioning

> The `tang` sections in `storage.luks` and `boot_device.luks` gained a new advertisement field. If specified, Ignition will use it to provision the Tang server binding rather than fetching the advertisement from the server at runtime. This allows the server to be unavailable at provisioning time. The advertisement can be obtained from the server with `curl http://tang.example.com/adv`.

Reviewer: ```suggestion
The `tang` sections in `storage.luks` and `boot_device.luks` gained a new `advertisement` field. If specified, Ignition will use it to provision the Tang server binding rather than fetching the advertisement from the server at runtime. This allows the server to be unavailable at provisioning time. The advertisement can be obtained from the server with `curl http://tang.example.com/adv`.
```

<!-- butane-config -->
```yaml
variant: fcos
version: 1.5.0
boot_device:
  luks:
    tang:
      - url: https://tang.example.com
        thumbprint: REPLACE-THIS-WITH-YOUR-TANG-THUMBPRINT
        advertisement: "{\"payload\": \"...\", \"protected\": \"...\", \"signature\": \"...\"}"
storage:
  luks:
    - name: luks-tang
      device: /dev/sdb
      clevis:
        tang:
          - url: https://tang.example.com
            thumbprint: REPLACE-THIS-WITH-YOUR-TANG-THUMBPRINT
            advertisement: "{\"payload\": \"...\", \"protected\": \"...\", \"signature\": \"...\"}"
```

### LUKS discard

> The `luks` section gained a new `discard` field. This can be specified in both the `storage` and/or the `boot_device`. If specified and true, the LUKS volume will issue discard commands to the underlying block device when blocks are freed. This improves performance and device longevity on SSDs and space utilization on thinly provisioned SAN devices, but leaks information about which disk blocks contain data.

Reviewer: Re the first two sentences, let's use language similar to the updated language in "Offline Tang provisioning".

<!-- butane-config -->
```yaml
variant: fcos
version: 1.5.0
grub:
  users:
    - name: admin
      password_hash: grub.pbkdf2.sha512.10000.874A958E5264...
```

### Offline Tang provisioning

The `tang` sections in `storage.luks` and `boot_device.luks` gained a new `advertisement` field. If specified, Ignition will use it to provision the Tang server binding rather than fetching the advertisement from the server at runtime. This allows the server to be unavailable at provisioning time. The advertisement can be obtained from the server with `curl http://tang.example.com/adv`.

<!-- butane-config -->
```yaml
variant: fcos
version: 1.5.0
boot_device:
  luks:
    tang:
      - url: https://tang.example.com
        thumbprint: REPLACE-THIS-WITH-YOUR-TANG-THUMBPRINT
        advertisement: "{\"payload\": \"...\", \"protected\": \"...\", \"signature\": \"...\"}"
storage:
  luks:
    - name: luks-tang
      device: /dev/sdb
      clevis:
        tang:
          - url: https://tang.example.com
            thumbprint: REPLACE-THIS-WITH-YOUR-TANG-THUMBPRINT
            advertisement: "{\"payload\": \"...\", \"protected\": \"...\", \"signature\": \"...\"}"
```

### LUKS discard

> The `luks` sections in `storage.luks` and `boot_device.luks` gained a new `discard` field. If specified and true, the LUKS volume will issue discard commands to the underlying block device when blocks are freed. This improves performance and device longevity on SSDs and space utilization on thinly provisioned SAN devices, but leaks information about which disk blocks contain data.

Reviewer: ```suggestion
The `luks` sections in `storage` and `boot_device` gained a new `discard` field. If specified and true, the LUKS volume will issue discard commands to the underlying block device when blocks are freed. This improves performance and device longevity on SSDs and space utilization on thinly provisioned SAN devices, but leaks information about which disk blocks contain data.
```

{% comment %}

## From Version 1.5.0 to Version 1.6.0-experimental

There are no breaking changes between versions 1.5.0 and 1.6.0-experimental of the `fcos` configuration specification. Any valid 1.5.0 configuration can be updated to a 1.6.0 configuration by changing the version string in the config.

## From Version 1.4.0 to Version 1.5.0

There are no breaking changes between versions 1.4.0 and 1.5.0 of the `fcos` configuration specification. Any valid 1.4.0 configuration can be updated to a 1.5.0 configuration by changing the version string in the config.

The following is a list of notable new features.
### Support Luks discard

> Discard; similarly to trim, improves performance for SSDs.

Reviewer: Please try to mimic the format of other entries in this file.  In many cases you can just reuse the text from the Ignition upgrade notes.

Author: Ah, yeah I missed the documentation section of the upgrade docs checklist. I see that now. Thank you

The following is a list of notable new features.

### Offline Tang provisioning

> The `tang` section gained a new `advertisement` field. If specified, Ignition will use it to provision the Tang server binding rather than fetching the advertisement from the server at runtime. This allows the server to be unavailable at provisioning time. The advertisement can be obtained from the server with `curl http://tang.example.com/adv`.

Reviewer: We'll also need a mention of `boot_device.luks.tang.advertisement`, and it should be included in the example.

Author: Gave it a swing!

<!-- butane-config -->
```yaml
variant: fcos
version: 1.5.0
storage:
  luks:
    - name: luks-tpm
      device: /dev/sdb
      open_options:
        - "--perf-no_read_workqueue"
        - "--perf-no_write_workqueue"
      clevis:
        tpm2: true
```

### Special mode bits supported

The `mode` field of the `files` and `directories` sections now respects the setuid, setgid, and sticky bits. Previous spec versions ignore these bits, but will generate a warning on Ignition ≥ 2.14.0.

<!-- butane-config -->
```yaml
variant: fcos
version: 1.5.0
storage:
  files:
    - path: /usr/local/bin/setuid
      mode: 2541
      contents:
        source: https://rootkit.example.com/setuid
  directories:
    - path: /var/local/tmp
      mode: 1023
```

### AWS S3 access point ARN support

The sections which allow fetching a remote URL now accept AWS S3 access point ARNs (`arn:aws:s3:<region>:<account>:accesspoint/<accesspoint>/object/<path>`) in the `source` field.

<!-- butane-config -->
```yaml
variant: fcos
version: 1.5.0
storage:
  files:
    - path: /etc/example
      mode: 420
      contents:
        source: arn:aws:s3:us-west-1:123456789012:accesspoint/test/object/some/path
```

> 

Reviewer: We should also document the new `grub` section.  It's sugar, so the Ignition migration guide doesn't have it.

Author: Ok, I will need to verse my self before I add an example.

Author: Ok, I looked at the existing example and tried to word it in a way that made sense. Though I suspect there might be feedback. LMK!

The following is a list of notable new features.

### Offline Tang provisioning

> The `tang` section gained a new `advertisement` field. This can be specified in both the `storage` and/or the `boot_device`. If specified, Ignition will use it to provision the Tang server binding rather than fetching the advertisement from the server at runtime. This allows the server to be unavailable at provisioning time. The advertisement can be obtained from the server with `curl http://tang.example.com/adv`.

Reviewer: Maybe "The `tang` sections in `storage.luks` and `boot_device.luks` gained a new `advertisement` field."?

<!-- butane-config -->
```yaml
variant: fcos
version: 1.5.0
boot_device:
  layout: ppc64le
  mirror:
    devices:
      - /dev/sda

>       - /dev/sdb

Reviewer: `layout` and `mirror` aren't relevant to these examples.  Let's remove, and revert the second disk to `sdb` instead of `sdc`.

  luks:
    tang:
      - url: https://tang.example.com
        thumbprint: REPLACE-THIS-WITH-YOUR-TANG-THUMBPRINT
        advertisement: "{\"payload\": \"...\", \"protected\": \"...\", \"signature\": \"...\"}"
storage:
  luks:
    - name: luks-tang
      device: /dev/sdc
      clevis:
        tang:
          - url: https://tang.example.com
            thumbprint: REPLACE-THIS-WITH-YOUR-TANG-THUMBPRINT
            advertisement: "{\"payload\": \"...\", \"protected\": \"...\", \"signature\": \"...\"}"
```

### LUKS discard

The `luks` section gained a new `discard` field. This can be specified in both the `storage` and/or the `boot_device`. If specified and true, the LUKS volume will issue discard commands to the underlying block device when blocks are freed. This improves performance and device longevity on SSDs and space utilization on thinly provisioned SAN devices, but leaks information about which disk blocks contain data.

<!-- butane-config -->
```yaml
variant: fcos
version: 1.5.0
boot_device:
  layout: ppc64le
  mirror:
    devices:
      - /dev/sda
      - /dev/sdb
  luks:

>       discard: true

Reviewer: Overindented.  Also, we need to configure some encryption here, probably `tpm2: true` for simplicity.

Reviewer: Not fixed.

Author: Sorry, I could have swore I fixed that...

Reviewer: Nope, the indentation hasn't changed.

storage:
  luks:
    - name: luks-tpm
      device: /dev/sdc
      discard: true
      clevis:
        tpm2: true
```

### LUKS open options

> The `luks` section gained a new `open_options` field. It is a list of options Ignition should pass to `cryptsetup luksOpen` when unlocking the volume. Ignition also passes `--persistent`, so any options that support persistence will be saved to the volume and automatically used for future unlocks. Any options that do not support persistence will only be applied to Ignition's initial unlock of the volume.

Reviewer: And for clarity, maybe "The `storage.luks` section gained..."?

<!-- butane-config -->
```yaml
variant: fcos
version: 1.5.0
storage:
  luks:
    - name: luks-tpm
      device: /dev/sdb
      open_options:
        - "--perf-no_read_workqueue"
        - "--perf-no_write_workqueue"
      clevis:
        tpm2: true
```

### Special mode bits supported

The `mode` field of the `files` and `directories` sections now respects the setuid, setgid, and sticky bits. Previous spec versions ignore these bits.

<!-- butane-config -->
```yaml
variant: fcos
version: 1.5.0
storage:
  files:
    - path: /usr/local/bin/setuid

>       mode: 0541

Reviewer: 0541 and 0123 don't include the setuid/setgid/sticky bits, and in fact aren't reasonable file modes.  Please use the original example values, converted to octal.

      contents:
        source: https://rootkit.example.com/setuid
  directories:
    - path: /var/local/tmp
      mode: 0123
```

### AWS S3 access point ARN support

The sections which allow fetching a remote URL now accept AWS S3 access point ARNs (`arn:aws:s3:<region>:<account>:accesspoint/<accesspoint>/object/<path>`) in the `source` field.

<!-- butane-config -->
```yaml
variant: fcos
version: 1.5.0
storage:
  files:
    - path: /etc/example

>       mode: 420

Reviewer: Octal mode.

The following is a list of notable new features.

### Offline Tang provisioning

The `tang` section gained a new `advertisement` field. If specified, Ignition will use it to provision the Tang server binding rather than fetching the advertisement from the server at runtime. This allows the server to be unavailable at provisioning time. The advertisement can be obtained from the server with `curl http://tang.example.com/adv`.

<!-- butane-config -->
```yaml
variant: fcos
version: 1.5.0
storage:
  luks:
    - name: luks-tang
      device: /dev/sdb
      clevis:
        tang:
          - url: https://tang.example.com
            thumbprint: REPLACE-THIS-WITH-YOUR-TANG-THUMBPRINT
            advertisement: "{\"payload\": \"...\", \"protected\": \"...\", \"signature\": \"...\"}"

>         tpm2: true

Author: Ah whoops, I went ahead and dropped it.

```

### LUKS discard

> The `luks` section gained a new `discard` field. If specified and true, the LUKS volume will issue discard commands to the underlying block device when blocks are freed. This improves performance and device longevity on SSDs and space utilization on thinly provisioned SAN devices, but leaks information about which disk blocks contain data.

Author: Ok, I gave it a swing.

```yaml
variant: fcos
version: 1.5.0
storage:
  luks:
    - name: luks-tpm
      device: /dev/sdb
      discard: true
      clevis:
        tpm2: true
```

### LUKS open options

The `luks` section gained a new `openOptions` field. It is a list of options Ignition should pass to `cryptsetup luksOpen` when unlocking the volume. Ignition also passes `--persistent`, so any options that support persistence will be saved to the volume and automatically used for future unlocks. Any options that do not support persistence will only be applied to Ignition's initial unlock of the volume.

<!-- butane-config -->
```yaml
variant: fcos
version: 1.5.0
storage:
  luks:
    - name: luks-tpm
      device: /dev/sdb
      open_options:
        - "--perf-no_read_workqueue"
        - "--perf-no_write_workqueue"
      clevis:
        tpm2: true
```

### Special mode bits supported

The `mode` field of the `files` and `directories` sections now respects the setuid, setgid, and sticky bits. Previous spec versions ignore these bits, but will generate a warning on Ignition ≥ 2.14.0.

<!-- butane-config -->
```yaml
variant: fcos
version: 1.5.0
storage:
  files:
    - path: /usr/local/bin/setuid

>       mode: 2541

Author: Oh ok, I changed out the leading number!
